core/src/saros/net/xmpp/contact/ContactStore.java

```diff
@@ -0,0 +1,199 @@
+package saros.net.xmpp.contact;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.jivesoftware.smack.Roster;
+import org.jivesoftware.smack.RosterEntry;
+import org.jivesoftware.smack.RosterGroup;
+import saros.net.xmpp.JID;
+
+/**
+ * Stores XMPPContacts and its group mappings.
+ *
+ * <p>A contact can be part of none, one or multiple groups.
+ *
+ * <p>All get methods are thread-safe, all others must be called within the {@link
+ * XMPPContactsService} Executor Thread.
+ */
+class ContactStore {
```

Comment:
Can you explain why all data must be cached ? It is because you want to mutate some states afterwards ?